Added a "Move to Folder" option to the file manager

mnakalay opened the pull request:
Deals with #5163 

As they are now, Folders in the file manager are difficult to manage. Sending files from one folder to another is impossible. Sending files that are not on page 1 to a folder which is on page 1 is impossible...

There are several situations where it is impossible to send files into a folder.

This pull request adds a "Move to Folder" option to the file contextual menu. It works for 1 or multi files.

mnakalay commented:
![dropdown](https://cloud.githubusercontent.com/assets/1488833/25507130/2e4ca1aa-2ba2-11e7-9921-3f48fb68898c.png)
![popup](https://cloud.githubusercontent.com/assets/1488833/25507133/316335fc-2ba2-11e7-82e1-28c255979d57.png)

mnakalay commented:
And here's a video showing how it works
[Move To Folder for the File Manager](https://www.youtube.com/watch?v=aXZIKme0IfE)

MrKarlDilkington commented:
This pull request addresses this issue:
 https://github.com/concrete5/concrete5/issues/5163

ramonleenders commented:
2 "issues" I'm seeing with this PR:

1. Page refresh takes too long to actually happen it seems. People can still do actions upon refreshing the page. Maybe add an overlay that prevents the user from clicking anywhere?
2. If folders are nested, it would be nice they would be displayed as nested within the "Move To" dialog that pops up. I could have multiple folders called "Test" and this way I don't know which one resides where. Also think of nested folders that are 3 or 4 deep. Clients can actually configure their folders this way, so we should tackle that I think.

Other than that, it's perfect :)

mlocati commented:
That's a great and very useful improvement. About refreshing the page: what if we simply remove the rows via javascript, without any further server call?

mnakalay commented:
@mlocati that would be a good solution, probably the most efficient one. But the core file manager js has a reloadFolder() function that si used in this kind of situation so I just modified the code to use that. It does call the server but that's how it's done when adding folders or dragging files in folders so I guess it's ok here as well.

mnakalay commented:
I'm still looking to address ramonleenders' other comment about nesting folders. I agree it would look better.

mnakalay commented:
I modified this to show folders hierarchically. [Here's an updated video](https://www.youtube.com/watch?v=orQJAuAPYh4)

Here's also an updated screenshot
![screenshot-c5-movetofolder-v2](https://cloud.githubusercontent.com/assets/1488833/25537897/5bd8d35a-2c38-11e7-8fbe-4537044de767.png)

ramonleenders commented:
Great job mate

aembler commented:
This is great. Really really nice job. My only questions would be regarding the addition of the multiple new functions in the Node class. Any reason why you went with those instead of just doing $node->populateChildren() and recursing through them using existing functions? I ask also because we might be able to remove a lot of the new code in favor of using existing code and simply integrating the the `Concrete\Core\File\Filesystem\FileFolderSelector::selectFileFolder()` method into your new dialog. (To be fair I don't think this method existed when you started working on this.)

Thoughts? I love the functionality and it is very much needed.

mnakalay commented:
Hi Andrew. Thanks for the nice words.
I didn't use populatechildren() because it doesn't filter by children type, it simply returns all children. Typically we would probably have a large number of children and a comparatively small number of folders. So my thinking was that getting all children to then maybe loop and filter by folders didn't make sense. The function I created filters by node type and is roughly copied on a function you have in the Page "helper". 

I also added a function to get all nodes of a specific type which I am not using. I just thought it would be a nice addition and could have its use.

If you feel it should be done differently, I'm ok to follow your instructions.

As fo the FileFolderSelector() function, I just checked and apparently it was already there and I didn't see it. Man I was so proud with my nicely indented and totally hacky list of folders....

So probably I don't need all this, I just need the selector?

How do you feel about still keeping the functions I added to Node.php? After all they could come in handy and you have the equivalent for pages, why not for nodes?

I'll wait to hear from you.

aembler commented:
I have to look this over again – which I will try to do next week.

mnakalay commented:
Hi @aembler Did you get a chance to have a look? should I replace the folders list with a selector?

aembler commented:
Thanks this is great. We merged it in and tweaked it a little bit.

mnakalay commented:
Hey Andrew, great to hear, can't wait to see your tweaks.
